python/pyarrow/tests/test_dataset.py

```diff
@@ -2672,47 +2672,57 @@ def test_feather_format(tempdir, dataset_reader):
         dataset_reader.to_table(ds.dataset(basedir, format="feather"))
 
 
-def _create_parquet_dataset_simple(root_path):
+def _create_parquet_dataset_simple(root_path, use_legacy_dataset=True):
     import pyarrow.parquet as pq
 
     metadata_collector = []
 
-    for i in range(4):
-        table = pa.table({'f1': [i] * 10, 'f2': np.random.randn(10)})
-        pq.write_to_dataset(
-            table, str(root_path), metadata_collector=metadata_collector
-        )
+    f1_vals = [item for chunk in range(4) for item in [chunk] * 10]
+    f2_vals = [item*10 for chunk in range(4) for item in [chunk] * 10]
+
+    table = pa.table({'f1': f1_vals, 'f2': f2_vals})
+    pq.write_to_dataset(
+        table, str(root_path), partition_cols=['f1'],
+        use_legacy_dataset=use_legacy_dataset,
+        metadata_collector=metadata_collector
+    )
```

Comment:
```suggestion
    root_path.mkdir()

    for i in range(4):
        table = pa.table({'f1': [i] * 10, 'f2': np.random.randn(10)})
        path = root_path / f"part-{i}.parquet"
        pq.write_table(table, str(path), metadata_collector=metadata_collector)
        # set the file path relative to the root of the partitioned dataset
        metadata_collector[-1].set_file_path(f"part-{i}.parquet")
```

Looking a bit more in detail, the goal of this helper method is actually to create a non-partitioned dataset (so just a directory with flat files), since the partitioned case is tested a bit more below with the helper function `_create_parquet_dataset_partitioned`. So my code suggestion here adapts it to use plain `write_table` (to overcome the issue that `write_to_dataset` generates identical files (and thus overwrites)

Reply:
Actually, this way it is of course not using the write_to_dataset non-legacy version, so if we are not doing that, we could also keep the original without passing the `use_legacy_dataset` keyword.

Reply:
But so there is a `_create_parquet_dataset_partitioned` a little bit below, and I think it is that one that should be parametrized with `use_legacy_dataset=True/False`

Reply:
I also ended up spinning myself around a few times here.  What I wanted is a test to ensure that we can round trip table -> _metadata -> factory -> dataset -> table using the metadata_collector.  However, changing use_legacy_dataset here fails (since the new dataset doesn't support append, a.k.a ARROW-12358).  An, both use_legacy_dataset versions fail when there is a partition because of ARROW-13269.

So I created a new, simple, test which tests the round trip with a single file and no append and no partitioning.  Tests for the other scenarios can be fleshed out when those issues are addressed.  I then restored all the old tests as they were.  Hopefully this works.